### ISSUE TITLE
Does not satisfy the constraints laid out in the rationale document

### ISSUE THREAD

**Issue by pzuraq** [mbrowne/bound-decorator#1]
Using an initializer to bind the method essentially converts the method into a field, which has all of the same problems as laid out in the rationale doc:

```js
class Counter extends React.Component {
    ...
    @bound
    handleClick() {
        this.setState({ count: this.state.count + 1 })
    }
    ...
}

class SpecialCounter extends Counter {
    handleClick() {
        console.log("SpecialCounter clicked")
    }
    ...
}

const specialCounter = new SpecialCounter()
// Which method gets called?
specialCounter.handleClick()
// (Note: this example is for illustrative purposes only; obviously we wouldn't really call handleClick() manually)
```

An additional constraint worth considering is that binding on both the superclass and the subclass should still Just Work ™️, even with calls to `super`:

```js
class Counter extends React.Component {
    ...
    @bound
    handleClick() {
        this.setState({ count: this.state.count + 1 })
    }
    ...
}

class SpecialCounter extends Counter {
    @bound
    handleClick() {
        super.handleClick();
        // do more things
    }
    ...
}
```

Currently, `super.handleClick()` will not exist on the superclass since it is a class field, and it will be overwritten by the time it gets to the child. This is also problematic for [the strategy of dynamically redefining the method within a getter](https://github.com/NoHomey/bind-decorator/blob/master/src/index.ts#L11), since the call to `super` will then overwrite the subclass's version.

The best strategy we have found to mitigate this is to use a `WeakMap` to map bound functions to the instance, and to bind the function in a getter:

```js
const BINDINGS_MAP = new WeakMap();

export function bound(desc) {
  let boundFn = desc.descriptor.value;

  desc.descriptor = {
    get() {
      let bindings = BINDINGS_MAP.get(this);

      if (bindings === undefined) {
        bindings = new Map();
        BINDINGS_MAP.set(this, bindings);
      }

      let fn = bindings.get(boundFn);

      if (fn === undefined) {
        fn = boundFn.bind(this);
        bindings.set(boundFn, fn);
      }

      return fn;
    }
  }

  return desc;
}
```

**Comment by mbrowne** [comment]
Thanks for the detailed report. There's actually a simpler solution to this problem that I will be including in the next version of this library. Currently the initializer is implemented as follows:

```js
    const { value } = descriptor
    function initializer() {
        return value.bind(this)
    }
```

I'm planning to change it to this:

```js
    function initializer() {
        return this[key].bind(this)
    }
```

which desugars to this:

```js
class Counter extends React.Component {
    constructor() {
        this.handleClick = this.handleClick.bind(this)
    }

    handleClick() {
        this.setState({ count: this.state.count + 1 })
    }
}

class SpecialCounter extends Counter {
    handleClick() {
        console.log('SpecialCounter clicked')
    }
}
```

I was confused by this at first as well, but this works because `this` is always evaluated in the context of the current instance, so even though `this.handleClick.bind(this)` is in the parent constructor, `this.handleClick` evaluates to `SpecialCounter.prototype.handleClick` since that takes precedence in the prototype chain.

BTW, this also solves another one of the examples in the rationale document, as explained in [this thread](https://github.com/tc39/proposal-decorators/issues/146#issuecomment-433582396).

The second example you mentioned actually works just fine with the current implementation. `super` still works because this implementation isn't just replacing the prototype method with a field...it creates a new field, yes, but also leaves the prototype method there as well. This alone is a major advantage over arrow functions.

There is another consideration regarding my planned implementation change...it should work with private methods as well. `this[key]` won't work with private methods so I'll need to check for those and bind them using the `descriptor.value` as the current implementation is doing. Babel doesn't yet support decorating private fields and methods (since the spec for that is still in flux) so I suppose I should go ahead and update the implementation for public fields without any additional checks for now.

BTW, I also had [an idea](https://github.com/tc39/proposal-decorators/issues/146#issuecomment-433385534) to allow this decorator to still work even if the prototype of a class is modified after it's created (as long as this prototype modification happens before you actually create the instance). But I think that extra complication probably isn't worth it for what is probably a pretty rare use case.

**Comment by pzuraq** [comment]
I believe the getter method would continue to work with private names, but that updated initializer method _would_ work better for mocking for sure, and it would possibly be a bit more performant. One nice part about the getter method is it makes the binding lazy - can't say how much of a benefit that would actually give an app at scale, I think binding is pretty cheap.

Can't believe we didn't consider that solution though, thanks for pointing it out, definitely going to consider benchmarking it and giving it a try 😄 

Maybe you could use the getter method for private fields, and the assignment method for anything public?

**Comment by mbrowne** [comment]
Fixed via https://github.com/mbrowne/bound-decorator/commit/d180f353b73b1a39a24e3dc3db561f04bd6b82ca.

I think this simple implementation should suffice for private methods, since inheritance isn't relevant for those.

**Comment by mbrowne** [comment]
FYI, I also published a new version of the npm package with this fix (v1.1.0)
